File: app/src/main/java/com/yenvth/soilDetectionApp/diction/DictionActivity.java
package com.yenvth.soilDetectionApp.diction;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatSpinner;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;


import com.yenvth.soilDetectionApp.R;
import com.yenvth.soilDetectionApp.models.DistrictModel;
import com.yenvth.soilDetectionApp.models.ProvinceModel;
import com.yenvth.soilDetectionApp.models.SoilModel;
import com.yenvth.soilDetectionApp.models.SoilParamModel;
import com.yenvth.soilDetectionApp.models.WardModel;
import com.yenvth.soilDetectionApp.soilDetail.SoilDetailActivity;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DictionActivity extends AppCompatActivity implements DictionView, View.OnClickListener, SoilAdapter.OnSoilItemClickListener {

    @BindView(R.id.btnBack)
    protected ImageView btnBack;
    @BindView(R.id.recycler_view_soil)
    protected RecyclerView recycler_view_soil;
    @BindView(R.id.spProvince)
    protected AppCompatSpinner spProvince;
    @BindView(R.id.spDistrict)
    protected AppCompatSpinner spDistrict;
    @BindView(R.id.spWard)
    protected AppCompatSpinner spWard;

    private SoilAdapter mAdapter;
    private ArrayList<SoilModel> dataList;
    private DictionPresenterImpl<DictionView> mPresenter;
    private String queryString;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_diction);
        getSupportActionBar().hide();
        ButterKnife.bind(this);
        mPresenter = new DictionPresenterImpl<>(DictionActivity.this, this);
        init();
        action();
    }

    private void init() {
        mPresenter.getListSoil(queryString, new SoilParamModel(0, 0, 0, 0, 0));
    }

    private void action() {
        btnBack.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btnBack:
                onBackPressed();
                break;
        }
    }

    @Override
    public void onGetListSoilSuccess(ArrayList<SoilModel> soilModels) {
        mAdapter = new SoilAdapter(this, soilModels, this);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
        recycler_view_soil.setLayoutManager(layoutManager);
        recycler_view_soil.setItemAnimator(new DefaultItemAnimator());
        recycler_view_soil.setAdapter(mAdapter);
    }

    @Override
    public void onGetListProvinceSuccess(ArrayList<ProvinceModel> provinceModels) {
        ArrayAdapter provinceAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, provinceModels);
        provinceAdapter.setDropDownViewResource(R.layout.item_spinner);
        spProvince.setAdapter(provinceAdapter);
    }

    @Override
    public void onGetListDistrictSuccess(ArrayList<DistrictModel> districtModels) {
        ArrayAdapter districtAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, districtModels);
        districtAdapter.setDropDownViewResource(R.layout.item_spinner);
        spDistrict.setAdapter(districtAdapter);
    }

    @Override
    public void onGetListWardSuccess(ArrayList<WardModel> wardModels) {
        ArrayAdapter wardAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, wardModels);
        wardAdapter.setDropDownViewResource(R.layout.item_spinner);
        spWard.setAdapter(wardAdapter);
    }

    @Override
    public void onSoilClickListener(SoilModel soilModel) {
        Intent intent = new Intent(this, SoilDetailActivity.class);
        intent.putExtra("soil_id", soilModel.getSoil_id());
        startActivity(intent);
    }
}

File: app/src/main/java/com/yenvth/soilDetectionApp/diction/DictionActivity.kt
package com.yenvth.soilDetectionApp.diction

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.View.OnFocusChangeListener
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.LayoutManager
import com.yenvth.soilDetectionApp.R
import com.yenvth.soilDetectionApp.databinding.ActivityDictionBinding
import com.yenvth.soilDetectionApp.room.entity.HistoryRecord
import com.yenvth.soilDetectionApp.soilDetail.SoilDetailActivity

class DictionActivity : AppCompatActivity(), View.OnClickListener {
    private val viewModel by viewModels<DictionViewModel>()
    private lateinit var binding: ActivityDictionBinding

    companion object {
        const val TAG = "DictionActivity"
    }

    private val mAdapter: SoilAdapter by lazy {
        SoilAdapter(this) { model ->
            viewModel.saveSearchHistory(
                HistoryRecord(
                    historyId = 0,
                    soilId = model.soilId,
                    System.currentTimeMillis()
                )
            )
            val intent = Intent(this, SoilDetailActivity::class.java)
            intent.putExtra("soilId", model.soilId)
            startActivity(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDictionBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.hide()
        init()
        action()
        setupObserve()
    }

    private fun init() {
        viewModel.getListSoil(queryString = "")

        val layoutManager: LayoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        binding.recyclerViewSoil.layoutManager = layoutManager
        binding.recyclerViewSoil.adapter = mAdapter
    }

    private fun action() {
        binding.btnBack.setOnClickListener(this)
        binding.btnSearch.setOnClickListener(this)
        binding.edSearch.onFocusChangeListener =
            OnFocusChangeListener { _, hasFocus ->
                if (hasFocus) {
                    binding.edSearch.setBackgroundResource(R.drawable.corner_search_edittext_select)
                } else {
                    binding.edSearch.setBackgroundResource(R.drawable.corner_search_edittext_unselect)
                }
            }
        binding.edSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                binding.btnDelete.visibility = if (s.isBlank()) View.GONE else View.VISIBLE
                Handler().postDelayed({ viewModel.getListSoil(s.toString().trim()) }, 100)
            }
        })
        binding.btnDelete.setOnClickListener {
            binding.edSearch.setText("")
        }
    }

    private fun setupObserve() {
        viewModel.soils.observe(this) { soilModels ->
            mAdapter.setData(soilModels)
        }

        viewModel.saveStatus.observe(this) {
            Log.d(TAG, "History saved")
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btnBack -> onBackPressed()
            R.id.btnSearch -> {
                val query = binding.edSearch.text.toString().trim()
                if (!TextUtils.isEmpty(query)) {
                    viewModel.getListSoil(query)
                }
            }
        }
    }
}